File: LC207_CourseSchedule/BFS/bfs.py
from collections import defaultdict, deque
from typing import List


class Solution:
    # BFS
    # Time Complexity: O(m+n)
    # Space Complexity: O(m+n)
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # edges[0] is all courses that require 0 as prerequisite
        edges = defaultdict(list)
        indegree = [0]*numCourses

        for pre in prerequisites:
            edges[pre[1]].append(pre[0])
            indegree[pre[0]] += 1
        
        queue = deque()
        # First take course from indegree = 0 (no prerequisite)
        for i in range(numCourses):
            if indegree[i] == 0:
                queue.append(i)
        
        visited = 0
        while len(queue) > 0:
            node = queue.popleft()
            visited += 1
            # Decrement in degree here
            for n in edges[node]:
                indegree[n] -= 1
                # Add courses with no rerequisite to the queue again
                if indegree[n] == 0:
                    queue.append(n)
        return visited == numCourses
        
########################### Test ######################
soln = Solution()
print(soln.canFinish(20, [[0,10],[3,18],[5,5],[6,11],[11,14],[13,1],[15,1],[17,4]]))  

File: LC207_CourseSchedule/BFS/bfs.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Solution {
    // BFS
    // Time Complexity: O(m+n)
    // Space Complexity: O(m+n)
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> inDegrees(numCourses, 0); // store the in degree for each node
        vector<vector<int>> neighbours(numCourses); // store nodes neighbour[i] points to

        // Initialize inDegrees and neighbour
        for(vector<int> pre : prerequisites){
            inDegrees[pre[0]]++;
            neighbours[pre[1]].push_back(pre[0]);
        }

        queue<int> queue;

        for(int i = 0; i < inDegrees.size(); i++){
            if(inDegrees[i] == 0){
                queue.push(i);
            }
        }

        int visited = 0;

        while(queue.size() > 0){
            int node = queue.front();
            queue.pop();
            visited++;
            for(int neighbour : neighbours[node]){
                // Decrement in degrees
                inDegrees[neighbour]--;

                if(inDegrees[neighbour] == 0){
                    queue.push(neighbour);
                }
            }
        }

        return numCourses == visited;
        
    }
};